### ISSUE TITLE
A little sneak peek for API Console v4

### ISSUE THREAD

**Issue by sichvoge** [mulesoft/api-console#328]
Hi all,

Now that the RAML Workgroup released 1.0, and we added support for it to our latest version of the API Console; we thought its time for us to share our next plans for the API Console with you. 

Version 4 will come with a complete face lifting for the UI that we think is more appropriated for an API documentation page, with easier navigation and a three column layout. Here a small image of what we have in the prototype.

![image](https://cloud.githubusercontent.com/assets/498114/18167367/1d5fce6e-7048-11e6-973a-6b20215f15c3.png)

We are also planning to add several new features such as JSON Schema visualization, improved syntax highlighting, code snippets, the grouping of resources, in-page search, and other general improvements to the user experience. 

We love our community, and also the people that build tools around API specification that helps consumers, or providers, to understand and test APIs. We already partner up with different vendors to create standards and to provide a consistent experience by building reusable / shareable components. Going forward, that will also be our strategy for the API Console. To achieve that, we teamed up with the [Advanced Rest Client](https://advancedrestclient.com/) to create reusable web components that can be easily combined like Lego blocks. We will take and combine some of these web components to construct the new version of the API Console.

As we are always keen to get early feedback from our user base, we build a prototype that shows you our current thinking and the chance to give us constructive feedback. Please follow [that link](https://rawgit.com/mulesoft/api-console/prototype/index.html), navigate through the example of an API documentation and let us know what you think here on this issue. Additionally, if you can spend some time, please visit the prototype for the full Advanced Rest Client [here](https://github.com/jarrodek/arc-prototype/tree/master/arc-prototype) and give feedback to the team their on what you think.

Our current plan is to release the new version probably in the first months of 2017, but that is subject to change. We will give you new updates as soon as possible of course.

What happens with the current version (v3)?

For now, we will continue supporting and maintaining it before we fully release version 4. After that, we could have two possibilities: 1) contribute the API Console v3 to the community and invite new contributors that will continue to support and enhance it, or 2) we put it into a branch so that people will still be able to use it. The first option is obviously preferable as in the second we will not do much maintenance anymore.

We really look forward to your feedback and thoughts!

Thanks,
The Mulesoft Team

**Comment by damimj** [comment]
About the code snippets, the center side of the page "jumps" because of the width of the different snippets. Since there's a lot of blank space on the left side, the snippet might be better if they have a fixed width. See gif below
![console-poc](https://cloud.githubusercontent.com/assets/9253676/18169252/98c47820-702f-11e6-881f-b9e2e3acf3bf.gif)

**Comment by sichvoge** [comment]
Thanks @damimj; makes a lot of sense, and very desired for the final version I agree.

**Comment by tehcyx** [comment]
It would be nice to have a visual divider between left handed navigation and main content. Or at least the grey active link box around "About" on load-up and then whatever is active after.

Is there any timeline for the prototype release of the "TRY IT" part?

**Comment by jarrodek** [comment]
Hi @tehcyx,
The prototype public test has begun today. See my blog post here: https://restforchrome.blogspot.com/2016/09/prototype-testing-invitation-to.html
It's not the same thing as API console but main components will be used to run both Advanced REST Client and API console. 
Consider spending some time testing the prototype and give us your feedback around design and functionality.

**Comment by sichvoge** [comment]
Dear community,

I am pleased to give you a short update on where we are with the new version of the API console. 

After the last few months, we made a lot of progress with the implementation as well as the UX. For the latter, we did several smaller UI research to improve the way how you will interact with the API console. 

I am really happy to announce that we will go into the last stage of our development circle - QA, and invite you to help us with that. We published the current version into the `release/4.0.0` branch including a guide on how you install and run it locally. Additionally, we hosted a version on http://mulesoft.github.io/api-console/ using our gh-pages branch. Please feel free to raise any problems on Github as an issue, but make sure to mention that you are using the `release/4.0.0` branch. If you want to provide feedback, please feel free to add a new comment on this issue.

We are planning to execute QA until the end of next week before we move forward. I will keep you updated.

Furthermore, we are working to extend your ability to use the API console in different scenarios. For example, you will be also able to use the console inside your pipeline using a command line interface including a way to use it inside your own scripts. Stay tuned. If there are other use cases you want us to tell about, please feel free to comment of course. 

Thanks in advance for all your help and we appreciate any time you spend helping us to test the new version. Also a huge round of applause for the team who made this happen.

We cannot wait to finally be able to merge it into the master. When that happens, we will put the previous 3.x.x version into a tag branch, so it will still be available for you.

**Comment by damimj** [comment]
Throws 592 JS errors and took 3 minutes to load the page. Just tested using Chrome.

**Comment by mafischer** [comment]
Can you publish this on npm with a beta tag?

**Comment by sichvoge** [comment]
@mafischer what are you doing with the npm package of the console today?

**Comment by mafischer** [comment]
I use it as a dependency for my web api and serve it as a static directory. It auto loads my web api's documentation at the /docs path using raml-cosnole-loader.

Here's an example https://github.com/mafischer/api-console-example

**Comment by sichvoge** [comment]
Finally released the [next version](https://github.com/mulesoft/api-console/releases/tag/v4.0.0) of the API Console. Please let us know your feedback.

**Comment by mafischer** [comment]
Any reason why you aren't publishing on npm?

**Comment by jarrodek** [comment]
@mafischer 
The API Console is build on top on web components specification. From the beginning we were thinking about using a package system that fit web components. At the time most popular for this architecture was bower (it still is but now we would choose Yarn over npm).
There's a lot of discussion around this over forums  and discussion groups but the fact is that the console has been created in bower (which is also supported by Yarn) and when installing the console you also installing almost 200 other dependencies. This currently can't be done with npm.

Even if you use npm as you primary package management system it doesn't limit you with using bower (or Yarn). It is basically one entry in your `package .json` script's map to install bower and then the console. Also, using our build tools for the console you can create production ready bundle inside the same same command. 

Consider this example:
```
"install-console": "npm install bower && node_modules/.bin/bower mulesoft/api-console"
```

This is all you need to have the console ready to work in your project. 

#### TLDR;
Architecturally, the API console is a shell application that consist with a number of independent web components. So when you are installing the console it also install all the dependencies. Many (if not most) aren't available on npm.
Also, web components are strictly client side code. That is not technically related to node. Sure, some of the packages you can run in browser but there's not many of them.

**Comment by jarrodek** [comment]
Also, please, check out our [build tools documentation](https://github.com/mulesoft/api-console/blob/master/docs/build-tools.md) to learn how to use nom mobiles or CLI tools to build production ready API Console.

**Comment by m31271n** [comment]
@jarrodek  Get it. But we can just release it on NPM with built content. Just like 3.x.x did.

**Comment by jarrodek** [comment]
@m31271n Not really. Right now you can build the console in 7 basic ways:
- standalone application
- embeddable element 
- with raml data source
- with parsed to JSON raml (to not to include parser into the build)
- as above but result of parsing being included into the build instead of separated file
- custom builds

It all depends on your project and how you'll use your console. Each of this options has it's pros and cons. I can't see how we could publish it on npm.
Easiest way for you would be use of our [npm module to build the console](https://github.com/mulesoft-labs/api-console-builder) (it downloads dependencies by default) or [CLI tool](https://github.com/mulesoft-labs/api-console-cli) that do the same.

**Comment by hadrienk** [comment]
@jarrodek How can I simply see the console while I am working on the RAML file with the new version? The build process is taking several minutes. With the previous version I was able to simply reload the page. Is something similar possible?

**Comment by jarrodek** [comment]
@hadrinek
I'm finalizing a tool that will allow you to do this. You can expect it very soon.
It is a node module that you can use directly in your project and it will be integrated with our CLI tool for the console.
I'll post a comment here when the module is ready and published. I'll be happy to get some feedback.
